feat(bundler): improve GitHub mirror URL generation with custom templates

thep0y opened the pull request:
- Add support for custom GitHub mirror URL templates via `TAURI_BUNDLER_TOOLS_GITHUB_MIRROR_TEMPLATE` env var
- Refactor URL generation logic into separate functions for better modularity
- Implement regex-based parsing of GitHub URLs to extract components
- Add unit tests for new URL generation functionality
- Remove direct dependency on `ureq::AgentBuilder` in favor of more flexible approach
- Update `create_agent_and_url` function to use new URL generation methods

Closes: https://github.com/tauri-apps/tauri/issues/10497

<!--
Before submitting a PR, please read https://github.com/tauri-apps/tauri/blob/dev/.github/CONTRIBUTING.md#pull-request-guidelines

1. Give the PR a descriptive title.

  Examples of good title:
    - fix(windows): fix race condition in event loop
    - docs: update example for `App::show`
    - feat: add `Window::set_fullscreen`

  Examples of bad title:
    - fix #7123
    - update docs
    - fix bugs

2. If there is a related issue, reference it in the PR text, e.g. closes #123.
3. If this change requires a new version, then add a change file in `.changes` directory with the appropriate bump, see https://github.com/tauri-apps/tauri/blob/dev/.changes/README.md
4. Ensure that all your commits are signed https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits
5. Ensure `cargo test` and `cargo clippy` passes.
6. Propose your changes as a draft PR if your work is still in progress.
-->

thep0y committed:
feat(bundler): improve GitHub mirror URL generation with custom templates

- Add support for custom GitHub mirror URL templates via `TAURI_BUNDLER_TOOLS_GITHUB_MIRROR_TEMPLATE` env var
- Refactor URL generation logic into separate functions for better modularity
- Implement regex-based parsing of GitHub URLs to extract components
- Add unit tests for new URL generation functionality
- Remove direct dependency on `ureq::AgentBuilder` in favor of more flexible approach
- Update `create_agent_and_url` function to use new URL generation methods

thep0y committed:
update `crates/tauri-cli/ENVIRONMENT_VARIABLES.md`

thep0y committed:
add file `.changes/bundler-github-mirror-template.md`

amrbashir commented on crates/tauri-bundler/src/bundle/windows/util.rs:
```diff
@@ -69,18 +69,58 @@ pub fn download_webview2_offline_installer(base_path: &Path, arch: &str) -> crat
+    mirror_url
+      .or(cdn_url)
+      .map(|alternative_url| (ureq::AgentBuilder::new().build(), alternative_url)),
+  )
```
This can be simplified into

```suggestion
  generate_mirror_url_from_template(url)
    .or_else(|_| generate_mirror_url_from_base(url))
    .map(|alt_url| {
      alt_url.map(|alt_url| (ureq::AgentBuilder::new().build(), alt_url))
    })
```

also notice usage of `or_else` instead of `or` which will lazily evaluate `generate_mirror_url_from_base` only if needed.

amrbashir commented on crates/tauri-bundler/src/bundle/windows/util.rs:
```diff
@@ -69,18 +69,58 @@ pub fn download_webview2_offline_installer(base_path: &Path, arch: &str) -> crat
+}
+
+fn create_agent_and_url(url: &str) -> crate::Result<(ureq::Agent, String)> {
+  generate_alternative_url(url)?.map_or_else(
+    || {
+      Ok((
+        ureq::AgentBuilder::new().try_proxy_from_env(true).build(),
+        url.to_owned(),
+      ))
+    },
+    Ok,
+  )
```
I generally dislike `map_or_else` since it is a badly named function and can be confusing at first glance, I would re-write this function without it
```suggestion
  generate_alternative_url(url).map(|alt_url| {
    alt_url.unwrap_or_else(|| (
        ureq::AgentBuilder::new().try_proxy_from_env(true).build(),
        url.to_owned(),
      ))
  })
```

thep0y replied:
This change is just a matter of personal coding preference and does not affect its functionality; I have no objection to this modification.

amrbashir replied:
correct, it is just a coding preference, think of it as a nit-pick but for consistency of the codebase this is a slightly important nit-pick.

amrbashir commented on crates/tauri-bundler/src/bundle/windows/util.rs:
```diff
@@ -196,3 +236,59 @@ pub fn os_bitness<'a>() -> Option<&'a str> {
+  use super::generate_mirror_url_from_template;
+  use std::env;
+
+  const URL: &str =
```
Let's call this `GOOD_URL`
```suggestion
  const GOOD_URL: &str =
```
and add another `BAD_URL` const next to it

thep0y replied:
`URL` is just a link to the source file that needs to be downloaded. I don't understand the point of changing it to `GOOD_URL`; this variable name has no relation to its function.

amrbashir replied:
it does in the context of test, from a first glance, it is not clear why this const is defined at all but in contrast, `GOOD_URL` and `BAD_URL` clearly states that first is for the good test case where we produce a new URL based on the template and the second for the bad case where we return None.

thep0y replied:
I still can't understand the purpose of naming it this way. The `URL` variable is just a resource address. If clarity in naming is needed, I think `GITHUB_ASSET_URL` would be more appropriate. I don't know where `GOOD_URL` and `BAD_URL` should be used.

amrbashir replied:
`URL` is only used in tests right? there is `test_generate_mirror_url_correctly1` and `test_generate_mirror_url_correctly2` that test whether we successfully got a URL from the template, so I'd consider these the "Good" cases. 

There is also `test_generate_mirror_url_no_env_var` and `test_generate_mirror_url_non_github_url` that cover cases when there is no env var or we are not downloading from github, in which I'd consider these the bad cases.

Maybe good and bad are not good names here so indeed `GITHUB_ASSET_URL` and `NON_GITHUB_ASSET_URL` (which is defined inline in `test_generate_mirror_url_non_github_url` test case) are better names.

thep0y committed:
Update crates/tauri-bundler/src/bundle/windows/util.rs

Co-authored-by: Amr Bashir <github@amrbashir.me>

thep0y committed:
Merge branch 'dev' into dev

thep0y committed:
Update crates/tauri-bundler/src/bundle/windows/util.rs

Co-authored-by: Amr Bashir <github@amrbashir.me>

thep0y committed:
Refactor map and unwrap_or_else in Windows bundler utilities, and update tests to use named constants for URLs

amrbashir commented on crates/tauri-bundler/src/bundle/windows/util.rs:
```diff
@@ -69,18 +69,52 @@ pub fn download_webview2_offline_installer(base_path: &Path, arch: &str) -> crat
+    })
+}
+
+fn generate_mirror_url_from_base(github_url: &str) -> crate::Result<Option<String>> {
```
I think it is fine to just return an `Option<String>` here

amrbashir commented on crates/tauri-bundler/src/bundle/windows/util.rs:
```diff
@@ -69,18 +69,52 @@ pub fn download_webview2_offline_installer(base_path: &Path, arch: &str) -> crat
+    return Ok(None);
   }
+
+  generate_mirror_url_from_template(url)
+    .or_else(|_| generate_mirror_url_from_base(url))
+    .map(|alt_url| alt_url.map(|alt_url| (ureq::AgentBuilder::new().build(), alt_url)))
```
looks like I provided you with wrong code, I thought `generate_mirror_url_from_template` returned a Result.

With the change requested above, the correct code should be 
```suggestion
  generate_mirror_url_from_template(url)
    .or_else(|| generate_mirror_url_from_base(url))
    .map(|alt_url| (ureq::AgentBuilder::new().build(), alt_url))
```

amrbashir commented on crates/tauri-bundler/src/bundle/windows/util.rs:
```diff
@@ -69,18 +69,52 @@ pub fn download_webview2_offline_installer(base_path: &Path, arch: &str) -> crat
+    .transpose()
+}
+
+fn generate_alternative_url(url: &str) -> crate::Result<Option<(ureq::Agent, String)>> {
```
With the change requested above, this can also be changed to just return an `Option<(ureq::Agent, Stirng)>`

thep0y committed:
Refactor Windows bundler utilities to simplify return types and error handling

- Change `generate_mirror_url_from_base` and `generate_alternative_url` to return `Option<T>` instead of `crate::Result<Option<T>>`.
- Simplify the logic in `create_agent_and_url` by removing the `crate::Result` wrapper.
- Remove unnecessary `Ok` and `?` usage, streamline error handling.

thep0y committed:
consolidate mirror URL generation tests

Combines two similar test functions into a single parameterized test for
better maintainability and easier addition of future test cases.

thep0y committed:
fix log statement placement in download function for Windows bundler

thep0y committed:
fix logging to use final URL after generating github mirror url in Windows bundler

thep0y committed:
Merge branch 'dev' into dev

lucasfernog commented on crates/tauri-bundler/src/bundle/windows/util.rs:
```diff
@@ -69,24 +69,53 @@ pub fn download_webview2_offline_installer(base_path: &Path, arch: &str) -> crat
-      AgentBuilder::new().try_proxy_from_env(true).build(),
-      url.to_owned(),
-    )),
+fn generate_mirror_url_from_template(github_url: &str) -> Option<String> {
```
I would name this to `generate_github_mirror_url_from_template` to be explicit that the URL must be a github URL

thep0y committed:
Merge branch 'tauri-apps:dev' into dev

thep0y committed:
Rename functions to clarify GitHub mirror URL generation

- Renamed `generate_mirror_url_from_template` to `generate_github_mirror_url_from_template`
- Renamed `generate_mirror_url_from_base` to `generate_github_mirror_url_from_base`
- Renamed `generate_alternative_url` to `generate_github_alternative_url`

amrbashir committed:
Merge branch 'dev' into dev

amrbashir committed:
use regex on linux as well

amrbashir commented on crates/tauri-cli/ENVIRONMENT_VARIABLES.md:
```diff
@@ -15,6 +15,7 @@ These environment variables are inputs to the CLI which may have an equivalent C
 - `TAURI_CLI_NO_DEV_SERVER_WAIT` — Skip waiting for the frontend dev server to start before building the tauri application.
 - `TAURI_LINUX_AYATANA_APPINDICATOR` — Set this var to `true` or `1` to force usage of `libayatana-appindicator` for system tray on Linux.
 - `TAURI_BUNDLER_WIX_FIPS_COMPLIANT` — Specify the bundler's WiX `FipsCompliant` option.
+- `TAURI_BUNDLER_TOOLS_GITHUB_MIRROR_TEMPLATE` - Specify a GitHub mirror template to download files and tools used by tauri bundler.
 - `TAURI_BUNDLER_TOOLS_GITHUB_MIRROR` - Specify a GitHub mirror to download files and tools used by tauri bundler.
```
```suggestion
- `TAURI_BUNDLER_TOOLS_GITHUB_MIRROR` - Specify a GitHub mirror to download files and tools used by tauri bundler.
- `TAURI_BUNDLER_TOOLS_GITHUB_MIRROR_TEMPLATE` - Specify a GitHub mirror template to download files and tools used by tauri bundler, for example: `https://mirror.example.com/<owner>/<repo>/releases/download/<version>/<asset>`.
```

amrbashir committed:
Update crates/tauri-cli/ENVIRONMENT_VARIABLES.md
